xdsl/dialects/fsm.py

```diff
@@ -363,6 +372,156 @@ def __init__(
+            result_types=[outputs],
+        )
+
+    def verify_(self):
+        if not isinstance(self.instance.owner, Instance):
+            raise VerifyException("The instance operand must be Instance")
+
+        # operand types must match the machine input types
+        # result types must match the machine output types
+        m = self.instance.owner.getMachine()
+
+        if m is None:
+            raise VerifyException("Machine definition does not exist.")
+
+        if not (
+            [operand.type for operand in self.inputs]
+            == [result for result in m.function_type.inputs]
+            and len(self.inputs) == len(m.function_type.inputs)
+        ):
+            raise VerifyException("Input types must be consistent with the machine's")
+
+        if not (
+            [operand.type for operand in self.outputs]
+            == [result for result in m.function_type.outputs]
+            and len(self.outputs) == len(m.function_type.outputs)
+        ):
+            raise VerifyException("Output types must be consistent with the machine's")
+
+
+@irdl_op_definition
+class HWInstance(IRDLOperation):
+    """Represents a hardware-style instance of a state machine,
+    including an instance name and a symbol reference of the machine. The inputs
+    and outputs are correponding to the inputs and outputs of the referenced
+    machine."""
+
+    name = "fsm.hw_instance"
+
+    sym_name = attr_def(StringAttr)
+    machine = attr_def(FlatSymbolRefAttr)
+    inputs = var_operand_def(AnyAttr())
+    clock = operand_def(signlessIntegerLike)
+    reset = operand_def(signlessIntegerLike)
+
+    outputs = var_result_def(AnyAttr())
+
+    def __init__(
+        self,
+        sym_name: str | StringAttr,
+        machine: str | FlatSymbolRefAttr,
+        inputs: Sequence[SSAValue | Operation],
+        clock: SSAValue | Operation,
+        reset: SSAValue | Operation,
+        outputs: Sequence[Attribute],
+    ):
+        if isinstance(machine, str):
+            machine = FlatSymbolRefAttr(machine)
+        clock = SSAValue.get(clock)
+        reset = SSAValue.get(reset)
+        attributes: dict[str, Attribute] = {}
+        if isinstance(sym_name, str):
+            sym_name = StringAttr(sym_name)
+        attributes["sym_name"] = sym_name
+        attributes["machine"] = machine
+        super().__init__(
+            operands=[inputs, clock, reset],
+            result_types=[outputs],
+            attributes=attributes,
+        )
+
+    def verify_(self):
+        m = SymbolTable.lookup_symbol(self, self.machine)
+        if isinstance(m, Machine):
+            if not (
+                [operand.type for operand in self.inputs]
+                == [result for result in m.function_type.inputs]
+                and len(self.inputs) == len(m.function_type.inputs)
+            ):
+                raise VerifyException(
+                    "Inputs types must be consistent with the machine's "
```

Comment:
Could you add the state machine's `sym_name` in these messages? Could it help debugging by knowing to which machine this error refers too based on the symbol table?

If so, adjust in the other similar cases where a symbol table lookup is involved too?